Implement file triggers 📂 (%transfiletriggerin)

cgwalters opened the pull request:
Depends: https://github.com/projectatomic/rpm-ostree/pull/868

File triggers are a post-RHEL7 thing; more information at
http://rpm.org/user_doc/file_triggers.html

There are two notable users I've been testing this with;
`glib2` and `vagrant`.  The `vagrant` one is more immediately urgent,
since it makes `vagrant-libvirt` work, which I currently rely on
for my workstation dev.

I've tested things successfully with `vagrant`, and I did verify that we run the
`glib2` ones when doing `rpm-ostree ex container`.

Long term, more transaction file triggers are likely to live in
"base" packages like `glib2`.  We don't implement those yet, but
extending this to do that shouldn't be too hard.

There was *significant* what I'd call reverse engineering of the
implementation in librpm.  The file triggers code there is spread out
and abstracted in a few different places in the code.  I found
trying to understand what header values were involved to be quite
tricky.

There are some corner cases like multiple patterns that I *think*
this does correctly, but could use more validation.  The main
question I had was - is it required that the patterns for e.g.
`%transfiletriggerin` and `%transfiletriggerun` be identical?

TODO: split out glibc fix, tests, remove debug prints

Closes: projectatomic#648

cgwalters commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -143,6 +153,13 @@ run_script_in_bwrap_container (int rootfs_fd,
+  if (strcmp (interp, "/bin/sh") == 0)
+    {
+      script_owned = g_strconcat ("set -x\n", script, NULL);
+      script = script_owned;
```
TODO; keep this but put under envvar for debugging?

peterbaouoft commented on src/libpriv/rpmostree-core.c:
```diff
@@ -2484,8 +2468,7 @@ run_pre_sync (RpmOstreeContext *self,
 
-  if (!rpmostree_pre_run_sync (pkg, hdr, self->ignore_scripts,
-                               rootfs_dfd, cancellable, error))
+  if (!rpmostree_pre_run_sync (pkg, hdr, rootfs_dfd, cancellable, error))
```
not sure if this would matter, but may be split into two lines like earlier implementation?

cgwalters commented:
Lifting WIP - this should be ready for review.  Fixes `vagrant-libvirt` for me on `fedora-27:fedora/rawhide/x86_64/workstation`.

cgwalters commented:
Note to self: update the commit message to say we now do handle base packages.

jlebon commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -317,23 +344,261 @@ run_script (const KnownRpmScriptKind *rpmscript,
+      while (buf->len > 0 && buf->str[buf->len-1] == '/')
+        g_string_truncate (buf, buf->len - 1);
+
+      guint this_pattern_matched = 0;
```
`n_pattern_matches`? And output var `out_n_matches`?

jlebon commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -317,23 +344,261 @@ run_script (const KnownRpmScriptKind *rpmscript,
+  return TRUE;
+}
+
+/* Given a set of file trigger @patterns (really subdirectories), traverse the
```
> (really subdirectories)

Is that how librpm treats them as well? Or do they do a simple prefix check on the full filename? The docs don't explicitly say they're directories.

jlebon commented on tests/common/libtest.sh:
```diff
@@ -423,6 +427,9 @@ $post
 ${posttrans:+%posttrans}
 $posttrans
 
+${transfiletriggerin:+%transfiletriggerin --} ${transfiletriggerin_pattern}
```
I'd make this:

```bash
${transfiletriggerin:+%transfiletriggerin -- ${transfiletriggerin_pattern}}
```

to be more explicit. (I guess that applies to `%post` as well).

jlebon commented on tests/vmcheck/test-layering-scripts.sh:
```diff
@@ -77,6 +77,25 @@ vm_cmd cat /usr/lib/prefixtest.txt > prefixtest.txt
+vm_build_rpm scriptpkg4 \
+             transfiletriggerin "/usr/bin" 'cat >/usr/share/transfiletriggerin-bin.txt'
+vm_build_rpm scriptpkg5 \
+             transfiletriggerin "/usr/lib" 'cat >/usr/share/transfiletriggerin-lib.txt'
```
We should have a third one as well which has both patterns to make sure we can handle a trigger with multiple patterns.

jlebon commented on tests/vmcheck/test-layering-scripts.sh:
```diff
@@ -77,6 +77,25 @@ vm_cmd cat /usr/lib/prefixtest.txt > prefixtest.txt
+             transfiletriggerin "/usr/bin" 'cat >/usr/share/transfiletriggerin-bin.txt'
+vm_build_rpm scriptpkg5 \
+             transfiletriggerin "/usr/lib" 'cat >/usr/share/transfiletriggerin-lib.txt'
+vm_build_rpm scriptpkg6
```
This looks like it isn't used below?

cgwalters replied:
Will add a comment.

jlebon commented on tests/vmcheck/test-layering-scripts.sh:
```diff
@@ -77,6 +77,25 @@ vm_cmd cat /usr/lib/prefixtest.txt > prefixtest.txt
+vm_rpmostree ex livefs
+vm_cmd cat /usr/share/transfiletriggerin-bin.txt > transfiletriggerin-bin.txt
+# Test that we actually got new files
+assert_file_has_content transfiletriggerin-bin.txt /usr/bin/scriptpkg5
```
Check for `/usr/bin/scriptpkg4` as well?

jlebon commented on tests/vmcheck/test-layering-scripts.sh:
```diff
@@ -77,6 +77,25 @@ vm_cmd cat /usr/lib/prefixtest.txt > prefixtest.txt
+vm_cmd cat /usr/share/transfiletriggerin-lib.txt | sort > transfiletriggerin-lib.txt
+vm_cmd find /usr/lib | sort > usrlib.txt
+diff -u usrlib.txt transfiletriggerin-lib.txt
+echo "ok transfiletriggerin"
```
How about also testing the following scenarios:

- pattern that doesn't match anything
- pkg with multiple `%transfiletriggerin` -- as opposed to `%post`, multiple file-based triggers are supported (most likely because it makes sense to split them by pattern)

cgwalters commented on configure.ac:
```diff
@@ -99,6 +88,13 @@ dnl bundled libdnf
+AS_IF([pkg-config --atleast-version=4.13.0.1 rpm], [
+AC_DEFINE([BUILDOPT_HAVE_RPMSQ_SET_INTERRUPT_SAFETY], 1, [Set to 1 if we have interrupt safety API])
+AC_DEFINE([BUILDOPT_HAVE_RPM_FILETRIGGERS], 1, [Set to 1 if we have file triggers])
+AC_DEFINE([BUILDOPT_HAVE_RPMFILES], 1, [Set to 1 if we have rpmfiles API]),
```
In the end it boils down to "built on RHEL" versus "built on Fedora" today which is pretty easy to keep track of.  If say OpenEmbedded started using rpm-ostree we'd be in a whole other feature matrix (including e.g. the rpm fork).  But I added a bit to the output.

cgwalters commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -317,23 +344,261 @@ run_script (const KnownRpmScriptKind *rpmscript,
+       * require /usr as a prefix too.
+       */
+      if (g_str_has_prefix (pattern, "usr/local") ||
+          !g_str_has_prefix (pattern, "usr/"))
```
AFAICS, none currently. From `rpm-specs-20170518.tar.xz`:
```
grep transfiletr *.spec
desktop-file-utils.spec:%transfiletriggerin -- %{_datadir}/applications
desktop-file-utils.spec:%transfiletriggerpostun -- %{_datadir}/applications
fontconfig.spec:%transfiletriggerin -- /usr/share/fonts /usr/share/X11/fonts/Type1 /usr/share/X11/fonts/TTF /usr/local/share/fonts
fontconfig.spec:%transfiletriggerpostun -- /usr/share/fonts /usr/share/X11/fonts/Type1 /usr/share/X11/fonts/TTF /usr/local/share/fonts
gap.spec:%transfiletriggerin -- %{gapdir}/pkg
gap.spec:%transfiletriggerpostun -- %{gapdir}/pkg
gdk-pixbuf2.spec:%transfiletriggerin -- %{_libdir}/gdk-pixbuf-2.0/2.10.0/loaders
gdk-pixbuf2.spec:%transfiletriggerpostun -- %{_libdir}/gdk-pixbuf-2.0/2.10.0/loaders
glib2.spec:%transfiletriggerin -- %{_libdir}/gio/modules
glib2.spec:%transfiletriggerpostun -- %{_libdir}/gio/modules
glib2.spec:%transfiletriggerin -- %{_datadir}/glib-2.0/schemas
glib2.spec:%transfiletriggerpostun -- %{_datadir}/glib-2.0/schemas
gtk2.spec:%transfiletriggerin -- %{_libdir}/gtk-2.0/immodules/ %{_libdir}/gtk-2.0/%{bin_version}/immodules/
gtk2.spec:%transfiletriggerpostun -- %{_libdir}/gtk-2.0/immodules/ %{_libdir}/gtk-2.0/%{bin_version}/immodules/
gtk3.spec:%transfiletriggerin -- %{_libdir}/gtk-3.0/3.0.0/immodules
gtk3.spec:%transfiletriggerpostun -- %{_libdir}/gtk-3.0/3.0.0/immodules
man-db.spec:%transfiletriggerin -- %{_mandir}
man-db.spec:%transfiletriggerpostun -- %{_mandir}
php-pear.spec:%transfiletriggerin -- %{pecl_xmldir}
php-pear.spec:%transfiletriggerun -- %{pecl_xmldir}
rpm.spec:- Fix invalid memory access on %transfiletriggerpostun (#1284645)
shared-mime-info.spec:%transfiletriggerin -- %{_datadir}/mime
shared-mime-info.spec:%transfiletriggerpostun -- %{_datadir}/mime
texlive.spec:%transfiletriggerin kpathsea -- %{_texdir}
texlive.spec:%transfiletriggerpostun kpathsea -- %{_texdir}
texlive.spec:%transfiletriggerin kpathsea -- %{_texdir}/texmf-dist/fonts/map/dvips/
texlive.spec:%transfiletriggerpostun kpathsea -- %{_texdir}/texmf-dist/fonts/map/dvips/
vagrant.spec:%transfiletriggerin -p %{_bindir}/ruby -- %{dirname:%{vagrant_plugin_spec}}
vagrant.spec:%transfiletriggerun -p %{_bindir}/ruby -- %{dirname:%{vagrant_plugin_spec}}
vim.spec:%transfiletriggerin common -- %{_datadir}/%{name}/vimfiles/doc
vim.spec:%transfiletriggerpostun common -- %{_datadir}/%{name}/vimfiles/doc
```

cgwalters commented:
OK, so this got rewritten again since I had an incomplete understanding of how librpm handles multiple patterns.  That took some more reverse engineering; it turns out there's a "triggers" array that maps the scripts to their patterns.

Updated the tests to have multiple patterns, also added "triggerun".

jlebon commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -317,23 +344,309 @@ run_script (const KnownRpmScriptKind *rpmscript,
+          n_total_matched += n_matched;
+        }
+
+      g_printerr ("matches: %u\n", n_total_matched);
```
Heh, left over from debugging?

jlebon commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -317,23 +344,309 @@ run_script (const KnownRpmScriptKind *rpmscript,
+                               GCancellable *cancellable,
+                               GError **error)
+{
+  GLNX_AUTO_PREFIX_ERROR("Finding matches", error);
```
So in this case here, we're not gaining anything from doing the prefixing in the callee right? (E.g. no computed strings, and only called from one place). In those cases, I think I'd still prefer prefixing in the caller. WDYT? (Overall, I still find it clearer to read than auto-prefixing).

jlebon replied:
Minor nit: space between function macro calls and opening parentheses.

cgwalters replied:
Hmm, but we're going to have both styles going forward right?  I don't see a clarity difference personally.

jlebon commented on src/libpriv/rpmostree-scripts.c:
```diff
@@ -317,23 +344,309 @@ run_script (const KnownRpmScriptKind *rpmscript,
+      !g_str_has_prefix (pattern, "usr/"))
+    return TRUE;
+
+  /* However, the printed buffer does have a leading / */
```
Minor nit: that "However" is misleading now -- I think the `strspn` used to be just above so it probably made sense then? (I thought this comment somehow related to the previous comment about fontconfig).

jlebon commented on tests/vmcheck/test-layering-scripts.sh:
```diff
@@ -77,6 +77,38 @@ vm_cmd cat /usr/lib/prefixtest.txt > prefixtest.txt
+license_combos="zlib-rpm systemd-tar-rpm sed-tzdata"
+license_un_combos="zlib systemd-rpm"
+vm_build_rpm scriptpkg4 \
+             transfiletriggerin "/usr/share/licenses/zlib /usr/share/licenses/rpm" 'cat >/usr/share/transfiletriggerin-license-zlib-rpm.txt' \
```
We can just `s/cat/sort` here right? (And the other ones) To avoid having to sort them down there.

cgwalters replied:
Fixup :arrow_down:

jlebon commented on tests/vmcheck/test-layering-scripts.sh:
```diff
@@ -77,6 +77,38 @@ vm_cmd cat /usr/lib/prefixtest.txt > prefixtest.txt
+    vm_cmd cat /usr/share/transfiletriggerin-license-${combo}.txt > transfiletriggerin-license-${combo}.txt
+    rm -f transfiletriggerin-fs-${combo}.txt
+    for path in $(echo ${combo} | sed -e 's,-, ,g'); do
+        vm_cmd find /usr/share/licenses/${path} -type f >> transfiletriggerin-fs-license-${combo}.txt
```
And here, just pipe through the host's `sort`.

jlebon commented:
@rh-atomic-bot r+ 1d07890

cgwalters commented:
Rebased :surfing_man:

jlebon commented:
@rh-atomic-bot r+ 03b072b

cgwalters commented:
Followup post to rpm-ecosystem about this: http://lists.rpm.org/pipermail/rpm-ecosystem/2017-July/000481.html
